beeai/supervisor/qe_data.py

```diff
@@ -0,0 +1,88 @@
+import asyncio
+from collections.abc import Callable
+from typing import Awaitable, Coroutine, TypeVar
+from pydantic import BaseModel
+
+from .http_utils import aiohttp_session
+
+QE_DATA_REPO = "https://gitlab.cee.redhat.com/otaylor/jotnar-qe-data"
+QE_DATA_URL = (
+    f"{QE_DATA_REPO}/-/raw/main/jotnar_qe_data.json?ref_type=heads&inline=false"
+)
+
+
+class TestLocationInfo(BaseModel):
+    component: str
+    qa_contact: str
+    tests_location: str | None = None
+    test_config_location: str | None = None
+    test_trigger_method: str | None = None
+    test_result_location: str | None = None
+    test_docs_url: str | None = None
+    notes: str | None = None
+
+
+R = TypeVar("R")
+
+
+def cache_async(
+    *,
+    max_age: float | None,
+) -> Callable[[Callable[[], Coroutine[None, None, R]]], Callable[[], Awaitable[R]]]:
+    """
+    Decorator to cache the result of an async function for a specified duration.
+    (For simplicity, max_age is required, and we don't need the standard decorator pattern
+    where it can be called with or without parentheses when no arguments are passed.)
+
+    Args:
+        func: The async function to be cached.
+        max_age: The maximum age (in seconds) for which the cached result is valid.
+    Returns:
+        An async function that returns the cached result if valid, otherwise calls the original function.
+    """
+
+    def decorator(
+        func: Callable[[], Coroutine[None, None, R]],
+    ) -> Callable[[], Awaitable[R]]:
+        _cache_task: asyncio.Task[R] | None = None
+        _cache_time: float | None = None
+
+        def wrapper() -> Awaitable[R]:
+            nonlocal _cache_task, _cache_time
+            now = asyncio.get_running_loop().time()
+
+            if (
+                _cache_task is None
+                or _cache_time is None
+                or (max_age is not None and (now - _cache_time) > max_age)
+            ):
+                _cache_task = asyncio.create_task(func())
+                _cache_time = now
```

Comment:
should probably add a `_cache_task.add_done_callback(...)` to drop from cache if the task is canceled or throws an exception. should time to live be from task start or task finishing? guess it doesn't really matter.

Reply:
Fooled around with far too many options for this (nerd-sniped!) - went with:

```diff
             if (
                 cache_task is None
+                or cache_task.cancelled()
+                or (cache_task.done() and cache_task.exception() is not None)
                 or cache_time is None
                 or (max_age is not None and (now - cache_time) > max_age)
             ):
                 cache_task = asyncio.create_task(func())
                 cache_time = now
```